File: src/engine_ts/decider.ts
/**
 * decider.ts — Deterministic State Machine Evaluator for TypeScript Engine.
 */

import { WorkflowInstance, WorkflowStatus, TaskInstance, TaskStatus, TaskDefinition } from "./types.js";

export interface DecisionResult {
  workflow_status: WorkflowStatus;
  tasks_to_schedule: TaskInstance[];
  tasks_to_retry: TaskInstance[];
  tasks_to_cancel: string[];
  stage_transitioned: boolean;
  new_stage_id?: string | null;
  is_terminal: boolean;
  failed_task?: TaskInstance | null;
  error_message?: string | null;
}

export class AgenticDecider {
  public evaluate(workflow: WorkflowInstance): DecisionResult {
    if (
      workflow.status === WorkflowStatus.COMPLETED ||
      workflow.status === WorkflowStatus.FAILED ||
      workflow.status === WorkflowStatus.CANCELLED
    ) {
      return {
        workflow_status: workflow.status,
        tasks_to_schedule: [],
        tasks_to_retry: [],
        tasks_to_cancel: [],
        stage_transitioned: false,
        is_terminal: true
      };
    }

    if (workflow.status === WorkflowStatus.PAUSED) {
      return {
        workflow_status: WorkflowStatus.PAUSED,
        tasks_to_schedule: [],
        tasks_to_retry: [],
        tasks_to_cancel: [],
        stage_transitioned: false,
        is_terminal: false
      };
    }

    const currentStage = workflow.workflow_def.stages[workflow.current_stage_index];
    if (!currentStage) {
      workflow.status = WorkflowStatus.COMPLETED;
      workflow.completed_at = Date.now();
      return {
        workflow_status: WorkflowStatus.COMPLETED,
        tasks_to_schedule: [],
        tasks_to_retry: [],
        tasks_to_cancel: [],
        stage_transitioned: false,
        is_terminal: true
      };
    }

    const scheduled: TaskInstance[] = [];
    const retries: TaskInstance[] = [];
    let allStageTasksDone = true;

    for (const taskDef of currentStage.tasks) {
      const taskInst = workflow.tasks[taskDef.id];

      if (!taskInst) {
        // Instantiate and schedule new task
        const newInst: TaskInstance = {
          task_id: taskDef.id,
          workflow_id: workflow.workflow_id,
          stage_id: currentStage.id,
          task_def: taskDef,
          status: TaskStatus.SCHEDULED,
          attempt: 1,
          input_data: { ...workflow.variables },
          output_data: {},
          worker_id: null,
          scheduled_at: Date.now(),
          trace_id: workflow.trace_id
        };
        workflow.tasks[taskDef.id] = newInst;
        scheduled.push(newInst);
        allStageTasksDone = false;
      } else if (
        taskInst.status === TaskStatus.SCHEDULED ||
        taskInst.status === TaskStatus.POLLED ||
        taskInst.status === TaskStatus.IN_PROGRESS ||
        taskInst.status === TaskStatus.GATE_EVALUATING ||
        taskInst.status === TaskStatus.DIAGNOSING
      ) {
        allStageTasksDone = false;
      } else if (taskInst.status === TaskStatus.FAILED) {
        const maxRetries = taskDef.retry_policy?.max_retries ?? 3;
        if (taskInst.attempt < maxRetries) {
          taskInst.attempt += 1;
          taskInst.status = TaskStatus.SCHEDULED;
          taskInst.worker_id = null;
          taskInst.lease_expires_at = null;
          retries.push(taskInst);
          allStageTasksDone = false;
        } else {
          // Check if Saga compensation is defined
          const compId = taskDef.compensation_task;
          if (compId && !workflow.tasks[compId]) {
            const compDef: TaskDefinition = {
              id: compId,
              type: "system.code",
              name: `Rollback compensation for ${taskInst.task_id}`,
              input_parameters: { code: "return { compensated: true };" }
            };
            const compInst: TaskInstance = {
              task_id: compId,
              workflow_id: workflow.workflow_id,
              stage_id: currentStage.id,
              task_def: compDef,
              status: TaskStatus.SCHEDULED,
              attempt: 1,
              input_data: { failed_task: taskInst.task_id, error: taskInst.error_message },
              output_data: {},
              scheduled_at: Date.now(),
              trace_id: workflow.trace_id
            };
            workflow.tasks[compId] = compInst;
            scheduled.push(compInst);
            allStageTasksDone = false;
          } else {
            // Terminal failure
            workflow.status = WorkflowStatus.FAILED;
            workflow.completed_at = Date.now();
            workflow.error_message = `Task ${taskInst.task_id} failed after ${taskInst.attempt} attempts: ${taskInst.error_message}`;
            return {
              workflow_status: WorkflowStatus.FAILED,
              tasks_to_schedule: [],
              tasks_to_retry: [],
              tasks_to_cancel: [],
              stage_transitioned: false,
              is_terminal: true,
              failed_task: taskInst,
              error_message: workflow.error_message
            };
          }
        }
      } else if (taskInst.status === TaskStatus.COMPLETED) {
        if (taskInst.output_data) {
          Object.assign(workflow.variables, taskInst.output_data);
          workflow.outputs[taskInst.task_id] = taskInst.output_data;
        }
      }
    }

    // Check if stage is complete
    if (allStageTasksDone && scheduled.length === 0 && retries.length === 0) {
      const nextStageIdx = workflow.current_stage_index + 1;
      if (nextStageIdx < workflow.workflow_def.stages.length) {
        workflow.current_stage_index = nextStageIdx;
        const nextStage = workflow.workflow_def.stages[nextStageIdx];
        const nextResult = this.evaluate(workflow);
        nextResult.stage_transitioned = true;
        nextResult.new_stage_id = nextStage.id;
        return nextResult;
      } else {
        workflow.status = WorkflowStatus.COMPLETED;
        workflow.completed_at = Date.now();
        return {
          workflow_status: WorkflowStatus.COMPLETED,
          tasks_to_schedule: [],
          tasks_to_retry: [],
          tasks_to_cancel: [],
          stage_transitioned: false,
          is_terminal: true
        };
      }
    }

    return {
      workflow_status: WorkflowStatus.RUNNING,
      tasks_to_schedule: scheduled,
      tasks_to_retry: retries,
      tasks_to_cancel: [],
      stage_transitioned: false,
      is_terminal: false
    };
  }
}

File: core/engine_py/decider.py
"""
decider.py — Deterministic State Machine & DAG Evaluator (The Decider).

Pure functional logic:
- Evaluates workflow DAG transitions without side effects
- Decides next tasks to schedule, retries to execute, or terminal statuses
- Enforces DNA inheritance: Stage 1 (Research) -> Stage 2 (Planning) -> Stage 3 (Implementation)
- Resolves parallel forks, join synchronization, and conditional switches
"""

import time
import uuid
from dataclasses import dataclass, field
from typing import Dict, List, Optional, Any, Tuple

from core.engine_py.models import (
    WorkflowInstance, WorkflowStatus, TaskInstance, TaskStatus, TaskDefinition
)


@dataclass
class DecisionResult:
    workflow_status: WorkflowStatus
    tasks_to_schedule: List[TaskInstance] = field(default_factory=list)
    tasks_to_retry: List[TaskInstance] = field(default_factory=list)
    tasks_to_cancel: List[str] = field(default_factory=list)
    stage_transitioned: bool = False
    new_stage_id: Optional[str] = None
    is_terminal: bool = False
    failed_task: Optional[TaskInstance] = None
    error_message: Optional[str] = None


class AgenticDecider:
    """State evaluator that calculates the next execution transitions for a workflow."""

    def evaluate(self, workflow: WorkflowInstance) -> DecisionResult:
        if workflow.status in [WorkflowStatus.COMPLETED, WorkflowStatus.FAILED, WorkflowStatus.CANCELLED]:
            return DecisionResult(workflow_status=workflow.status, is_terminal=True)

        if workflow.status == WorkflowStatus.PAUSED:
            return DecisionResult(workflow_status=WorkflowStatus.PAUSED)

        current_stage = workflow.get_current_stage()
        if not current_stage:
            # All stages completed!
            workflow.status = WorkflowStatus.COMPLETED
            workflow.completed_at = time.time()
            return DecisionResult(workflow_status=WorkflowStatus.COMPLETED, is_terminal=True)

        # 1. Inspect existing tasks in the current stage
        stage_task_defs = {t.id: t for t in current_stage.tasks}
        scheduled: List[TaskInstance] = []
        retries: List[TaskInstance] = []

        all_stage_tasks_done = True

        for task_def in current_stage.tasks:
            task_inst = workflow.tasks.get(task_def.id)

            if not task_inst:
                # Task not yet instantiated or scheduled
                new_inst = TaskInstance(
                    task_id=task_def.id,
                    workflow_id=workflow.workflow_id,
                    stage_id=current_stage.id,
                    task_def=task_def,
                    status=TaskStatus.SCHEDULED,
                    trace_id=workflow.trace_id,
                    input_data=dict(workflow.variables)
                )
                workflow.tasks[task_def.id] = new_inst
                scheduled.append(new_inst)
                all_stage_tasks_done = False

            elif task_inst.status in [TaskStatus.SCHEDULED, TaskStatus.POLLED, TaskStatus.IN_PROGRESS, TaskStatus.GATE_EVALUATING, TaskStatus.DIAGNOSING]:
                all_stage_tasks_done = False

            elif task_inst.status == TaskStatus.FAILED:
                # Check retry policy
                max_retries = task_def.retry_policy.max_retries
                if task_inst.attempt < max_retries:
                    task_inst.attempt += 1
                    task_inst.status = TaskStatus.SCHEDULED
                    task_inst.worker_id = None
                    task_inst.lease_expires_at = None
                    retries.append(task_inst)
                    all_stage_tasks_done = False
                else:
                    # Check if a Saga compensation task is defined
                    comp_id = task_def.compensation_task
                    if comp_id and comp_id not in workflow.tasks:
                        comp_def = TaskDefinition(
                            id=comp_id,
                            type="system.code",
                            name=f"Rollback compensation for {task_inst.task_id}",
                            input_parameters={"code": "result = 'COMPENSATED'"}
                        )
                        comp_inst = TaskInstance(
                            task_id=comp_id,
                            workflow_id=workflow.workflow_id,
                            stage_id=current_stage.id,
                            task_def=comp_def,
                            status=TaskStatus.SCHEDULED,
                            trace_id=workflow.trace_id,
                            input_data={"failed_task": task_inst.task_id, "error": task_inst.error_message}
                        )
                        workflow.tasks[comp_id] = comp_inst
                        scheduled.append(comp_inst)
                        all_stage_tasks_done = False
                    else:
                        # Irrecoverable task failure
                        workflow.status = WorkflowStatus.FAILED
                        workflow.completed_at = time.time()
                        workflow.error_message = f"Task {task_inst.task_id} failed after {task_inst.attempt} attempts: {task_inst.error_message}"
                        return DecisionResult(
                            workflow_status=WorkflowStatus.FAILED,
                            is_terminal=True,
                            failed_task=task_inst,
                            error_message=workflow.error_message
                        )

            elif task_inst.status == TaskStatus.COMPLETED:
                # Merge outputs into workflow variables
                if task_inst.output_data:
                    workflow.variables.update(task_inst.output_data)
                    workflow.outputs[task_inst.task_id] = task_inst.output_data

        # 2. Check if current stage has completed
        if all_stage_tasks_done and not scheduled and not retries:
            next_stage_idx = workflow.current_stage_index + 1
            if next_stage_idx < len(workflow.workflow_def.stages):
                workflow.current_stage_index = next_stage_idx
                next_stage = workflow.get_current_stage()
                # Recursively evaluate the new stage to schedule its initial tasks
                recurse_res = self.evaluate(workflow)
                recurse_res.stage_transitioned = True
                recurse_res.new_stage_id = next_stage.id if next_stage else None
                return recurse_res
            else:
                # All stages successfully finished!
                workflow.status = WorkflowStatus.COMPLETED
                workflow.completed_at = time.time()
                return DecisionResult(workflow_status=WorkflowStatus.COMPLETED, is_terminal=True)

        return DecisionResult(
            workflow_status=WorkflowStatus.RUNNING,
            tasks_to_schedule=scheduled,
            tasks_to_retry=retries
        )
